Implement celery task to award program certs.

jimabramson opened the pull request:
ECOM-3354

jimabramson commented:
@zubair-arbi @rlucioni please review.

There is one outstanding item to do here, which is to get the correct service user in place for making the credentials API call.  Everything else is finished (except for changes based on your feedback).

ahsan-ul-haq commented on lms/djangoapps/certificates/api.py:
```diff
@@ -116,7 +125,7 @@ def generate_user_certificates(student, course_key, course=None, insecure=False,
         forced_grade=forced_grade
     )
-    if cert.status in [CertificateStatuses.generating, CertificateStatuses.downloadable]:
+    if is_passing_status(cert.status):
```
Why not use `CertificateStatuses.is_passing_status(cert.status)` here and avoid method above?

rlucioni commented on openedx/core/djangoapps/programs/tasks.py:
```diff
@@ -4,12 +4,112 @@
+        list of program ids
+
+    """
+    return client.programs.complete.post(course_certificates)
```
The program completion endpoint returns [an object](https://github.com/edx/programs/blob/master/programs/apps/api/v1/views.py#L130), not a list. This will end up returning a dictionary as-is.

rlucioni commented on openedx/core/djangoapps/programs/tasks.py:
```diff
@@ -24,32 +124,35 @@ def award_program_certificates(username):
+                username,
+                program_id,
+            )
+            award_program_certificate(credentials_client, student, program_id)
```
The student's username is provided as an argument to this task. I'd pass that along here instead of pulling the username off the User object.

jimabramson replied:
i was doing that to be consistent with other helper methods in this module which need the User object, but i suppose the fewer places we do that, the better.  will change.

rlucioni commented on openedx/core/djangoapps/programs/tasks.py:
```diff
@@ -4,12 +4,112 @@
+        client:
+            programs API client (EdxRestApiClient)
+        course_certificates:
+            iterable of dicts with structure {'course_id': course_key, 'mode': cert_type}
```
The program completion endpoint expects [an object](https://github.com/edx/programs/blob/master/programs/apps/api/v1/views.py#L62-L69), not a flat iterable.

rlucioni commented on openedx/core/djangoapps/programs/tests/test_tasks.py:
```diff
@@ -0,0 +1,244 @@
+            body='[1, 2, 3]',
+            content_type='application/json',
+        )
+        payload = [
```
This payload should be updated to match what programs expects. Or, we should update programs so that it knows what to do with this payload.

rlucioni commented on openedx/core/djangoapps/programs/tests/test_tasks.py:
```diff
@@ -0,0 +1,244 @@
+        httpretty.register_uri(
+            httpretty.POST,
+            'http://test-server/programs/complete/',
+            body='[1, 2, 3]',
```
This body should be updated to match what programs returns. Or, we should update programs so that it returns a body of this form.

rlucioni commented on openedx/core/djangoapps/programs/tests/test_tasks.py:
```diff
@@ -0,0 +1,244 @@
+
+from openedx.core.djangoapps.credentials.tests.mixins import CredentialsApiConfigMixin
+from openedx.core.djangoapps.programs.tests.mixins import ProgramsApiConfigMixin
+from openedx.core.djangoapps.programs import tasks
```
Nitpick: I think the most recent version of Pylint would flag this block of imports. I would separate standard library, third-party, and first-party imports into separate groups.

jimabramson replied:
i've occasionally seen (and used) the approach of further subdividing first-party imports to separate the package under test from other supporting packages.   i guess that may be something i dreamed / invented rather than an actual practice, so i'm changing this as requested.

rlucioni commented on openedx/core/djangoapps/programs/tests/test_tasks.py:
```diff
@@ -0,0 +1,244 @@
+@mock.patch('openedx.core.djangoapps.programs.tasks.award_program_certificate')
+@mock.patch('openedx.core.djangoapps.programs.tasks.get_awarded_certificate_programs')
+@mock.patch('openedx.core.djangoapps.programs.tasks.get_completed_programs')
+@mock.patch('openedx.core.djangoapps.programs.tasks.get_completed_courses')
```
I'm curious about why you chose to structure the tests the way you have. Since you're testing the helper functions independently above, I understand why you're mocking them here. However, I'm wondering if things would be cleaner if httpretty was used instead. I think doing so would allow you to recycle setup methods and might make this easier to maintain as the code changes.

jimabramson replied:
it may be for lack of imagination on my part but i found httpretty kind of difficult to work with especially with multiple API calls.  That's part of why I have things structured this way.  I was running into a lot of trouble debugging httpretty when it didn't seem to be catching the uris i was registering and this was taking a lot of time.

I feel like part of the value of using the slumber client is that we can trust it to translate method/function calls into HTTP interactions, which is another reason that I think httpretty, while very useful in various situations, isn't absolutely essential for validating correct usage of the APIs.  But I'm open to changing if your argument is that coverage/confidence will improve if we switch to doing that.

rlucioni replied:
I like that httpretty lets us run more of our code during testing, as opposed to mocking out significant chunks. I also like that httpretty doesn't pin us to a specific implementation. For example, the helper functions tested here might change in the future, forcing a developer to revisit these mocks. It may not be worth fighting to get httpretty working here, but I do think it has the potential to make for cleaner code.

rlucioni commented on openedx/core/djangoapps/programs/tests/test_tasks.py:
```diff
@@ -0,0 +1,244 @@
+        ]
+        mock_get_completed_courses.return_value = completed_courses
+
+        tasks.award_program_certificates(self.student.username)
```
I'm surprised this isn't failing. Network calls to the programs and credentials APIs aren't being mocked, and the mocks you're creating for the helper methods don't have any return value assigned.

jimabramson replied:
The main concern of this test is that the call we make to programs contains the correct payload, specifically the result from `get_completed_courses`.  The return values from the other mocked helpers are not interesting, but I am validating that based on a certain set of completed courses, we will ask Programs the right question about completion.

rlucioni replied:
I'm on board with you, but, for example, why doesn't the call to Programs fail? You patch `get_completed_programs` but don't provide a return value for it, so I'd expect the code to call through to the actual implementation and fail when it tries to make a network call. I'm just trying to work out what's going on here.

jimabramson replied:
Patches return a generic Mock object when you don't set up any other return value (as opposed to falling back to actual implementation).  The task function generally doesn't operate on any of the return values from these helpers, just passes them on to the next helper, which is why none of the task code is complaining about whats actually inside those return values.

rlucioni replied:
Got it, thank you.

rlucioni commented on openedx/core/djangoapps/programs/tasks.py:
```diff
@@ -24,32 +124,35 @@ def award_program_certificates(username):
-        LOGGER.debug('calling credentials service to issue certificate for user %s in program %s', username, program_id)
-        # TODO
+    if new_program_ids:
+        LOGGER.debug('generating new program certificates for %s in programs: %r', username, new_program_ids)
```
Would info-level logging me more appropriate here?

jimabramson replied:
I put these in for (ahem) debugging purposes but I think you're right - would probably make sense to promote some to INFO.   Will adjust.

rlucioni commented:
@jimabramson do we have control over what queue these tasks are routed to? Routing these tasks to a named queue will make this process easier to monitor.

rlucioni commented:
@jimabramson along the same lines: should this task be versioned? If the task signature changes at some point, versioning would help us prevent workers running the old task from picking up new tasks.

jimabramson commented:
@rlucioni could you link me an example of how you are specifying queues in your recent async work?  I'll follow that model.

regarding versioning, yeah, good point, i'll put something there too.

rlucioni commented:
@jimabramson here's how I specify queues for ecommerce: https://github.com/edx/ecommerce/blob/master/ecommerce/settings/base.py#L463

I use directory structure to implement task versioning. You can refer to ecommerce worker for an example: https://github.com/edx/ecommerce-worker/tree/master/ecommerce_worker/fulfillment.

jimabramson commented:
@ahsan-ul-haq @rlucioni pushed a fixup addressing feedback so far, except:
* specifying queue for task
* specifying task version
* service user / authentication

I'm also planning to push up some improvements to error handling; I will keep the fixup commits separate to make reviewing easier.

jimabramson commented:
@rlucioni @ahsan-ul-haq @zubair-arbi ready for re-review.

@feanil will ping offline with some questions about this.

rlucioni commented on lms/djangoapps/certificates/api.py:
```diff
@@ -34,6 +34,15 @@
 log = logging.getLogger("edx.certificate")
 
 
+def is_passing_status(cert_status):
```
Any reason not to get rid of this and use your class method directly?

jimabramson replied:
i deliberately added this so that we wouldn't be importing models in other packages, and instead use the api.py functions.   this is consistent with the current usage/intent of the api.py module.

rlucioni replied:
I see. What's wrong with importing models across packages? I've heard this given as a justification for api.py before, but haven't questioned it.

If we're going to stick with api.py, shouldn't this function be used in [lms/djangoapps/instructor_task/tasks_helper.py](https://github.com/edx/edx-platform/pull/11341/files#diff-9b7fcb710c553d15f0d138a19ec082d0R1451)?

jimabramson replied:
initially i planned to update that, but that module breaks the encapsulation convention in far more egregious ways (because it predates it) and would need much more work to bring into conformance.  this gets pretty far afield of the intent of the current PR.

rlucioni replied:
Okay, thanks. I was just curious.

rlucioni commented on openedx/core/djangoapps/programs/tasks/v1/tasks.py:
```diff
@@ -0,0 +1,220 @@
+    return [
+        {'course_id': cert['course_key'], 'mode': cert['type']}
+        for cert in all_certs
+        if is_passing_status(cert['status'])
```
This looks like the only place `CertificateStatuses.is_passing_status` isn't called directly.

jimabramson replied:
see above

rlucioni commented on openedx/core/djangoapps/programs/signals.py:
```diff
@@ -47,5 +47,5 @@ class of the object instance that sent this signal
     # import here, because signal is registered at startup, but items in tasks are not yet able to be loaded
-    from openedx.core.djangoapps.programs import tasks
-    tasks.award_program_certificates.delay(user.username)
+    from openedx.core.djangoapps.programs.tasks.v1.tasks import award_program_certificates
```
I've found that a nice pattern is to give the directory containing your versioned tasks the same name you would give a queue holding the tasks. For example, instead of `openedx.core.djangoapps.programs.tasks.v1.tasks`, I might use something like `openedx.core.djangoapps.programs.certification.v1.tasks`. What do you think?

jimabramson replied:
i'm not feeling it.  If we add tasks related to other programs use cases, this would mean a second separate directory tree with identical structure.  Also, I don't want to muddy the waters around how/why we are versioning subpackages; `tasks.vN` keeps that pretty clear.

rlucioni replied:
Sounds good.

rlucioni commented on openedx/core/djangoapps/programs/tasks/v1/tasks.py:
```diff
@@ -0,0 +1,220 @@
+
+    except Exception, exc:  # pylint: disable=broad-except
+        LOGGER.exception('Failed to determine program certificates to be awarded for user %s', username)
+        raise self.retry(exc=exc)
```
Should there be a retry policy defined for this task? As it stands, the task will be retried immediately a total of 3 times (Celery's default) before failing. Would exponential backoff be helpful here? Are 3 retries sufficient?

jimabramson replied:
good catch.  should definitely be a longer delay on retries.  most likely causes of errors are transient issues reaching APIs or misconfigurations, and a longer delay will increase the odds of a clean recovery (allowing more time to intervene and fix).  also, this operation is relatively relaxed in urgency, as there's no user waiting anxiously for a result.

My gut instinct is to do 7 retries with exp backoff, starting at 1 minute.  So the last retry delay would be around 2h before we gave up.   @feanil any thoughts or comments on this?

jimabramson replied:
@rlucioni are you comfortable with implementing whatever we decide for custom retry in a follow-up PR?  I'd like to get the core functionality merged ASAP so that integration-related work isn't blocked.

rlucioni replied:
@jimabramson yes, implementing retry logic separately separately sounds good. For context, the ecomworker uses exponential backoff based on the number of retries attempted; see [here](https://github.com/edx/ecommerce-worker/blob/master/ecommerce_worker/fulfillment/v1/tasks.py#L53). By default, the worker is configured to retry 11 times. 11 retries with exponential backoff yields a maximum waiting time of 2047 seconds (about 30 minutes).

rlucioni commented on openedx/core/djangoapps/programs/tasks/v1/tests/test_tasks.py:
```diff
@@ -0,0 +1,426 @@
+        self.assertFalse(mock_get_completed_courses.called)
+        self.assertFalse(mock_get_completed_programs.called)
+        self.assertFalse(mock_get_awarded_certificate_programs.called)
+        self.assertFalse(mock_award_program_certificate.called)
```
You might consider doing this a little more DRYly as:
```python
for mocked_helper in (
    mock_get_completed_courses,
    mock_get_completed_programs,
    mock_get_awarded_certificate_programs,
    mock_award_program_certificate
):
    self.assertFalse(getattr(mocked_helper, 'called'))
```

jimabramson replied:
@rlucioni i added something DRYer than that, take a look.

rlucioni commented on openedx/core/djangoapps/programs/tasks/v1/tests/test_tasks.py:
```diff
@@ -0,0 +1,410 @@
+        with mock.patch(TASKS_MODULE + '.LOGGER.warning') as mock_warning:
+            tasks.award_program_certificates.delay(self.student.username).get()
+            self.assertTrue(mock_warning.called)
+        for mock_helper in mock_helpers:
```
So dry! Nice.

rlucioni commented:
:+1:

ahsan-ul-haq commented on lms/djangoapps/instructor_task/tasks_helper.py:
```diff
@@ -1448,7 +1448,7 @@ def generate_students_certificates(
         )
 
-        if status in [CertificateStatuses.generating, CertificateStatuses.downloadable]:
+        if CertificateStatuses.is_passing_status(status):
```
Why not use the `api` method you implemented here as well?

jimabramson replied:
I answered the same question from Renzo up here https://github.com/edx/edx-platform/pull/11341#discussion_r51930027 - basically, that module uses certificates models all over the place (instead of the api), and I didn't want to add a larger refactoring of that module into this PR.

ahsan-ul-haq replied:
Oh I missed that sounds good.

ahsan-ul-haq commented on openedx/core/djangoapps/programs/tasks/v1/tasks.py:
```diff
@@ -0,0 +1,220 @@
+
+LOGGER = get_task_logger(__name__)
+
+SERVICE_USERNAME = 'credentials_service_user'  # FIXME
```
Could you please put this in settings so that if we want to use this anywhere else we could use it from settings?

jimabramson replied:
good call, done

ahsan-ul-haq commented on openedx/core/djangoapps/programs/tasks/v1/tasks.py:
```diff
@@ -0,0 +1,220 @@
+
+    """
+    id_token = get_id_token(student, api_config.OAUTH2_CLIENT_NAME)
+    return EdxRestApiClient(api_config.internal_api_url, jwt=id_token)
```
It looks like api util method so I would suggest to move this in [here](https://github.com/edx/edx-platform/blob/master/openedx/core/lib/edx_api_utils.py) and if you could please change [this](https://ahsan-credentials.sandbox.edx.org/).

jimabramson replied:
i agree - i know it seems small but I've moved this to a separate ticket ECOM-3652.  I think there may be a few other opportunities to improve reuse, but need to get this merged.

ahsan-ul-haq replied:
I saw that I am happy to get this merged and work on the reusability part later.

ahsan-ul-haq commented on openedx/core/djangoapps/programs/tasks/v1/tasks.py:
```diff
@@ -0,0 +1,220 @@
+        list of program ids
+
+    """
+    return client.programs.complete.post({'completed_courses': course_certificates})['program_ids']
```
Why not have all utility methods in [utils.py](https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/programs/utils.py)

jimabramson replied:
my reason for not doing this is that all of these helper methods were broken out of the task logic in order to make it more testable.  The methods are only meant for use in the context of this task which is why I don't think they need to go into another shared place, at least not until we have a need for them elsewhere.

ahsan-ul-haq commented:
A few suggestions and questions other than that it looks good to me :+1:
